Add `input_format_binary_max_type_complexity` setting to protect against maliciously crafted binary type encodings

rienath opened the pull request:
Closes https://github.com/ClickHouse/ClickHouse/issues/92354. 

Introduce a new setting `input_format_binary_max_type_complexity` (default: 1000) that limits the total number of type nodes that can be decoded in binary format. For example, `Map(String, UInt32)` counts as 3 nodes. When the limit is exceeded, decoding fails fast with a clear error message rather than consuming resources. Setting to 0 disables the limit.

### Changelog category (leave one):
- Improvement


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Introduce a new setting `input_format_binary_max_type_complexity` that limits the total number of type nodes that can be decoded in binary format to prevent malicious payloads.

rienath committed:
Add input_format_binary_max_type_complexity setting

rienath committed:
Merge branch 'master' into binary-encoding-nodes

rienath committed:
Fix wrong setting name

alexey-milovidov commented on src/DataTypes/DataTypesBinaryEncoding.cpp:
```diff
@@ -607,9 +622,12 @@ String encodeDataType(const DataTypePtr & type)
-DataTypePtr decodeDataType(ReadBuffer & buf)
+DataTypePtr decodeDataTypeImpl(ReadBuffer & buf, size_t & complexity)
 {
-    checkStackSize();
```
Just in case, we can leave `checkStackSize();`, but as it is heavy, we can use something like

```
if (complexity % 128 == 0)
    checkStackSize();
```

rienath replied:
Good idea, I will add it! We can have a scenario where stack size is within limits, but the input still takes a long time to process (i.e. `Tuple(Tuple(UInt8 × 999999) × 10000)`)

alexey-milovidov commented:
Also, how about adding `checkStackSize` in `DB::IDataType::getDefaultSerialization`?

rienath committed:
Make test more dramatic

Previous test was getting the needed behaviour, but it did not demonstrate the malicious scenario. The new test would time out without the introduced changes

rienath committed:
Refactor

rienath committed:
Respect stack size

rienath commented:
Added

rienath committed:
Merge remote-tracking branch 'origin/master' into binary-encoding-nodes

rienath committed:
Bump setting from 25.12 to 26.1

Avogar reviewed (approved):
LGTM

rienath committed:
Merge branch 'master' into binary-encoding-nodes
